File: clients/android/app/src/main/java/eu/mycellium/android/MessengerViewModel.kt
package eu.mycellium.android

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import uniffi.mycellium_mobile.ClientState
import uniffi.mycellium_mobile.ContactInfo
import uniffi.mycellium_mobile.ContactSecurityInfo
import uniffi.mycellium_mobile.ConversationInfo
import uniffi.mycellium_mobile.DeliveryState
import uniffi.mycellium_mobile.EventKind
import uniffi.mycellium_mobile.MessageInfo
import uniffi.mycellium_mobile.MobileClient
import uniffi.mycellium_mobile.ProfileInfo

data class MessengerUiState(
    val initialized: Boolean = false,
    val busy: Boolean = false,
    val clientState: ClientState = ClientState.NEEDS_LOGIN,
    val loginRequested: Boolean = false,
    val error: String? = null,
    val notice: String? = null,
    val conversations: List<ConversationInfo> = emptyList(),
    val contacts: List<ContactInfo> = emptyList(),
    val profile: ProfileInfo? = null,
    val pendingCount: ULong = 0u,
    val selectedUserId: String? = null,
    val selectedTitle: String = "",
    val messages: List<MessageInfo> = emptyList(),
    val security: ContactSecurityInfo? = null,
    val startupError: String? = null,
)

class MessengerViewModel(application: Application) : AndroidViewModel(application) {
    private companion object {
        const val TAG = "Mycellium"
    }

    private val identityStore = AndroidIdentityStore(application)
    private val mutableState = MutableStateFlow(MessengerUiState())
    val state: StateFlow<MessengerUiState> = mutableState.asStateFlow()
    private var client: MobileClient? = null

    init {
        viewModelScope.launch(Dispatchers.IO) {
            runCatching {
                val dataDir = application.filesDir.resolve("mycellium").absolutePath
                val opened = MobileClient.open(dataDir, identityStore.load(), null)
                client = opened
                refreshAll()
                opened
            }.onSuccess {
                mutableState.update { it.copy(initialized = true, startupError = null) }
                pollEvents()
            }.onFailure { error ->
                Log.e(TAG, "mobile client failed to open", error)
                client = null
                showError(error)
            }
        }
    }

    fun requestLogin(email: String) = action {
        require(email.isNotBlank()) { "Enter your email address" }
        requireClient().requestEmailLogin(email.trim())
        mutableState.update {
            it.copy(loginRequested = true, notice = "Check your email for the login code")
        }
    }

    fun confirmLogin(code: String) = action {
        require(code.isNotBlank()) { "Enter the code from your email" }
        val result = requireClient().confirmEmailLogin(code.trim())
        result.identitySecret?.let { secret ->
            try {
                identityStore.save(secret)
            } finally {
                secret.fill(0)
            }
        }
        mutableState.update {
            it.copy(
                loginRequested = false,
                notice = if (result.created) "Account created" else "This device is active",
            )
        }
        refreshAll()
    }

    fun confirmLoginLink(link: String) = action {
        val result = requireClient().confirmEmailLoginLink(link)
        result.identitySecret?.let { secret ->
            try {
                identityStore.save(secret)
            } finally {
                secret.fill(0)
            }
        }
        mutableState.update {
            it.copy(
                loginRequested = false,
                notice = if (result.created) "Account created" else "This device is active",
            )
        }
        refreshAll()
    }

    fun saveProfile(handle: String, displayName: String) = action {
        requireClient().saveProfile(handle, displayName)
        mutableState.update { it.copy(notice = "Profile saved") }
        refreshAll()
    }

    fun addContact(card: String, nickname: String) = action {
        require(card.isNotBlank()) { "Paste a connection card" }
        val contact = requireClient().addContact(card.trim(), nickname.trim().ifEmpty { null })
        mutableState.update { it.copy(notice = "${contact.nickname} added") }
        refreshAll()
    }

    fun removeContact(contact: ContactInfo) = action {
        requireClient().removeContact(contact.nickname)
        mutableState.update {
            it.copy(security = null, selectedUserId = null, notice = "Person removed")
        }
        refreshAll()
    }

    fun openConversation(userId: String, title: String) = action {
        mutableState.update {
            it.copy(selectedUserId = userId, selectedTitle = title, security = null)
        }
        refreshMessages()
    }

    fun closeConversation() {
        mutableState.update { it.copy(selectedUserId = null, selectedTitle = "", messages = emptyList()) }
    }

    fun sendMessage(text: String) = action {
        val userId = mutableState.value.selectedUserId ?: error("Choose a conversation")
        val result = requireClient().sendText(userId, text)
        mutableState.update {
            it.copy(
                notice = if (result == DeliveryState.DELIVERED) {
                    "Delivered"
                } else {
                    "Direct connection unavailable. This device will keep trying."
                },
            )
        }
        refreshAll()
        refreshMessages()
    }

    fun retryPending() = action {
        val waiting = requireClient().retryPending()
        mutableState.update {
            it.copy(notice = if (waiting == 0uL) "Pending messages delivered" else "$waiting still pending")
        }
        refreshAll()
    }

    fun showSecurity(userId: String) = action {
        val security = requireClient().contactSecurity(userId)
        mutableState.update { it.copy(security = security) }
    }

    fun dismissSecurity() {
        mutableState.update { it.copy(security = null) }
    }

    fun verifyContact(userId: String) = action {
        requireClient().verifyContact(userId)
        mutableState.update { it.copy(notice = "Identity verified", security = null) }
        refreshAll()
    }

    fun acceptIdentityChange(userId: String) = action {
        requireClient().acceptIdentityChange(userId)
        mutableState.update { it.copy(notice = "New identity accepted", security = null) }
        refreshAll()
    }

    fun setContactBlocked(userId: String, blocked: Boolean) = action {
        val current = requireClient()
        current.setContactBlocked(userId, blocked)
        val security = current.contactSecurity(userId)
        mutableState.update {
            it.copy(
                notice = if (blocked) "Person blocked" else "Person unblocked",
                security = security,
            )
        }
    }

    fun clearMessage() {
        mutableState.update { it.copy(error = null, notice = null) }
    }

    fun restartLogin() {
        mutableState.update { it.copy(loginRequested = false, error = null, notice = null) }
    }

    fun onForeground() {
        if (client == null) return
        action {
            runCatching { requireClient().refreshConnectivity() }
            requireClient().refreshDeviceStatus()
            refreshAll()
        }
    }

    private fun action(block: suspend () -> Unit) {
        if (client == null) {
            mutableState.update {
                it.copy(error = it.startupError ?: "Mycellium is still starting")
            }
            return
        }
        if (mutableState.value.busy) {
            return
        }
        mutableState.update { it.copy(busy = true, error = null, notice = null) }
        viewModelScope.launch(Dispatchers.IO) {
            runCatching { block() }
                .onFailure { error ->
                    Log.e(TAG, "action failed", error)
                    showError(error)
                }
            mutableState.update { it.copy(busy = false) }
        }
    }

    private suspend fun refreshAll() {
        val current = requireClient()
        val state = current.state()
        val ready = state == ClientState.READY || state == ClientState.REPLACED
        val profile = if (ready) current.profile() else null
        val conversations = if (ready) current.conversations() else emptyList()
        val contacts = if (ready) current.contacts() else emptyList()
        val pending = if (ready) current.pendingCount() else 0u
        mutableState.update {
            it.copy(
                clientState = state,
                profile = profile,
                conversations = conversations,
                contacts = contacts,
                pendingCount = pending,
            )
        }
    }

    private suspend fun refreshMessages() {
        val userId = mutableState.value.selectedUserId ?: return
        val messages = requireClient().messages(userId)
        mutableState.update { it.copy(messages = messages) }
    }

    private suspend fun pollEvents() {
        while (true) {
            delay(1_000)
            val events = runCatching { requireClient().pollEvents() }.getOrDefault(emptyList())
            if (events.isNotEmpty()) {
                val latest = events.last()
                mutableState.update {
                    it.copy(
                        notice = if (latest.kind == EventKind.ERROR) null else latest.message,
                        error = if (latest.kind == EventKind.ERROR) latest.message else null,
                    )
                }
                runCatching {
                    refreshAll()
                    refreshMessages()
                }
            }
        }
    }

    private fun requireClient(): MobileClient = client ?: error("Mycellium is still starting")

    private fun showError(error: Throwable) {
        val startupError = if (client == null) error.message ?: "Could not open this device" else null
        mutableState.update {
            it.copy(
                initialized = true,
                busy = false,
                startupError = startupError,
                error = error.message ?: "Something went wrong",
                notice = null,
            )
        }
    }
}

File: clients/apple/App/MessengerViewModel.swift
import Foundation

struct MessengerState {
    var initialized = false
    var busy = false
    var clientState: ClientState = .needsLogin
    var loginRequested = false
    var error: String?
    var notice: String?
    var conversations: [ConversationInfo] = []
    var contacts: [ContactInfo] = []
    var profile: ProfileInfo?
    var pendingCount: UInt64 = 0
    var selectedUserId: String?
    var selectedTitle = ""
    var messages: [MessageInfo] = []
    var security: ContactSecurityInfo?
}

private struct ClientSnapshot: Sendable {
    let clientState: ClientState
    let conversations: [ConversationInfo]
    let contacts: [ContactInfo]
    let profile: ProfileInfo?
    let pendingCount: UInt64
}

@MainActor
final class MessengerViewModel: ObservableObject {
    @Published private(set) var state = MessengerState()
    private var client: MobileClient?
    private var eventTask: Task<Void, Never>?

    init() {
        Task { await bootstrap() }
    }

    deinit { eventTask?.cancel() }

    func requestLogin(email: String) {
        guard !email.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            fail("Enter your email address")
            return
        }
        perform { client in
            _ = try client.requestEmailLogin(email: email.trimmingCharacters(in: .whitespacesAndNewlines))
            await MainActor.run {
                self.state.loginRequested = true
                self.state.notice = "Check your email for the login code"
            }
        }
    }

    func confirmLogin(code: String) {
        guard !code.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            fail("Enter the code from your email")
            return
        }
        perform { client in
            let result = try client.confirmEmailLogin(code: code.trimmingCharacters(in: .whitespacesAndNewlines))
            if var secret = result.identitySecret {
                defer { secret.resetBytes(in: 0..<secret.count) }
                try KeychainIdentityStore().save(secret)
            }
            await MainActor.run {
                self.state.loginRequested = false
                self.state.notice = result.created ? "Account created" : "This device is active"
            }
            try await self.refresh(client)
        }
    }

    func confirmLogin(link: URL) {
        perform { client in
            let result = try client.confirmEmailLoginLink(link: link.absoluteString)
            if var secret = result.identitySecret {
                defer { secret.resetBytes(in: 0..<secret.count) }
                try KeychainIdentityStore().save(secret)
            }
            await MainActor.run {
                self.state.loginRequested = false
                self.state.notice = result.created ? "Account created" : "This device is active"
            }
            try await self.refresh(client)
        }
    }

    func saveProfile(handle: String, displayName: String) {
        perform { client in
            _ = try client.saveProfile(handle: handle, displayName: displayName)
            await MainActor.run { self.state.notice = "Profile saved" }
            try await self.refresh(client)
        }
    }

    func addContact(card: String, nickname: String) {
        perform { client in
            let cleanName = nickname.trimmingCharacters(in: .whitespacesAndNewlines)
            let contact = try client.addContact(
                connectionCard: card.trimmingCharacters(in: .whitespacesAndNewlines),
                nickname: cleanName.isEmpty ? nil : cleanName
            )
            await MainActor.run { self.state.notice = "\(contact.nickname) added" }
            try await self.refresh(client)
        }
    }

    func removeContact(_ contact: ContactInfo) {
        perform { client in
            try client.removeContact(nickname: contact.nickname)
            await MainActor.run {
                self.state.security = nil
                self.state.selectedUserId = nil
                self.state.notice = "Person removed"
            }
            try await self.refresh(client)
        }
    }

    func openConversation(userId: String, title: String) {
        state.selectedUserId = userId
        state.selectedTitle = title
        state.security = nil
        refreshMessages()
    }

    func closeConversation() {
        state.selectedUserId = nil
        state.selectedTitle = ""
        state.messages = []
    }

    func sendMessage(_ text: String) {
        guard let userId = state.selectedUserId else { return }
        perform { client in
            let delivery = try client.sendText(userId: userId, text: text)
            await MainActor.run {
                self.state.notice = delivery == .delivered
                    ? "Delivered"
                    : "Direct connection unavailable. This device will keep trying."
            }
            try await self.refresh(client)
            try await self.refreshMessages(client)
        }
    }

    func retryPending() {
        perform { client in
            let waiting = try client.retryPending()
            await MainActor.run {
                self.state.notice = waiting == 0 ? "Pending messages delivered" : "\(waiting) still pending"
            }
            try await self.refresh(client)
        }
    }

    func showSecurity(userId: String) {
        perform { client in
            let security = try client.contactSecurity(userId: userId)
            await MainActor.run { self.state.security = security }
        }
    }

    func dismissSecurity() { state.security = nil }

    func verifyContact(userId: String) {
        perform { client in
            try client.verifyContact(userId: userId)
            await MainActor.run {
                self.state.security = nil
                self.state.notice = "Identity verified"
            }
            try await self.refresh(client)
        }
    }

    func acceptIdentityChange(userId: String) {
        perform { client in
            try client.acceptIdentityChange(userId: userId)
            await MainActor.run {
                self.state.security = nil
                self.state.notice = "New identity accepted"
            }
            try await self.refresh(client)
        }
    }

    func setContactBlocked(userId: String, blocked: Bool) {
        perform { client in
            try client.setContactBlocked(userId: userId, blocked: blocked)
            let security = try client.contactSecurity(userId: userId)
            await MainActor.run {
                self.state.security = security
                self.state.notice = blocked ? "Person blocked" : "Person unblocked"
            }
        }
    }

    func restartLogin() {
        state.loginRequested = false
        state.error = nil
        state.notice = nil
    }

    func clearBanner() {
        state.error = nil
        state.notice = nil
    }

    func onForeground() {
        guard let client else { return }
        Task {
            _ = try? await Task.detached { try client.refreshConnectivity() }.value
            _ = try? await Task.detached { try client.refreshDeviceStatus() }.value
            try? await refresh(client)
        }
    }

    private func bootstrap() async {
        do {
            let opened = try await Task.detached {
                let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
                let directory = base.appendingPathComponent("Mycellium", isDirectory: true)
                try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
                var values = URLResourceValues()
                values.isExcludedFromBackup = true
                var mutableDirectory = directory
                try mutableDirectory.setResourceValues(values)
                let secret = try KeychainIdentityStore().load()
                return try MobileClient.open(
                    dataDir: directory.path,
                    identitySecret: secret,
                    registryUrl: nil
                )
            }.value
            client = opened
            try await refresh(opened)
            state.initialized = true
            startEventPolling(opened)
        } catch {
            fail(error.localizedDescription)
            state.initialized = true
        }
    }

    private func perform(
        _ operation: @escaping @Sendable (MobileClient) async throws -> Void
    ) {
        guard !state.busy, let client else { return }
        state.busy = true
        state.error = nil
        state.notice = nil
        Task {
            do {
                try await Task.detached { try await operation(client) }.value
            } catch {
                fail(error.localizedDescription)
            }
            state.busy = false
        }
    }

    private func refresh(_ client: MobileClient) async throws {
        let snapshot = try await Task.detached {
            let clientState = client.state()
            let available = clientState == .ready || clientState == .replaced
            return try ClientSnapshot(
                clientState: clientState,
                conversations: available ? client.conversations() : [],
                contacts: available ? client.contacts() : [],
                profile: available ? client.profile() : nil,
                pendingCount: available ? client.pendingCount() : 0
            )
        }.value
        state.clientState = snapshot.clientState
        state.conversations = snapshot.conversations
        state.contacts = snapshot.contacts.sorted {
            $0.nickname.localizedCaseInsensitiveCompare($1.nickname) == .orderedAscending
        }
        state.profile = snapshot.profile
        state.pendingCount = snapshot.pendingCount
    }

    private func refreshMessages() {
        guard let client else { return }
        Task { try? await refreshMessages(client) }
    }

    private func refreshMessages(_ client: MobileClient) async throws {
        guard let userId = state.selectedUserId else { return }
        let messages = try await Task.detached { try client.messages(userId: userId) }.value
        guard state.selectedUserId == userId else { return }
        state.messages = messages
    }

    private func startEventPolling(_ client: MobileClient) {
        eventTask?.cancel()
        eventTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(1))
                let events = await Task.detached { client.pollEvents() }.value
                guard let latest = events.last else { continue }
                if latest.kind == .error {
                    state.error = latest.message
                } else {
                    state.notice = latest.message
                }
                try? await refresh(client)
                try? await refreshMessages(client)
            }
        }
    }

    private func fail(_ message: String) {
        state.error = message
        state.notice = nil
        state.busy = false
    }
}
